File: haskell/grade-school/School.hs
module School( sorted, add, empty, grade, School ) where

import qualified Data.Map.Strict as Map
import Data.List

type Names = [String]
type Grade = Int
type School = Map.Map Grade Names

empty :: School
empty = Map.empty

add :: Grade -> String -> School -> School
add grade = Map.insertWith ( (sort .) . (++) ) grade . (:[])

grade :: Grade -> School -> Names
grade = Map.findWithDefault []

sorted :: School -> [ (Grade, Names) ]
sorted = Map.toAscList

File: scala/grade-school/src/main/scala/School.scala

class School {
  type DB = Map[Int, Seq[String]]

  private var store: DB = Map.empty

  def add(name: String, g: Int) =
    this.store = store.updated(g, store.getOrElse(g, Seq.empty) :+ name)


  def db: DB = store

  def grade(g: Int): Seq[String] = store.getOrElse(g, Seq.empty)

  def sorted: DB =
    store.toList.sortBy(_._1).map{ case (k: Int, v: Seq[String])=> (k, v.sorted) }.toMap
}
